### ISSUE TITLE
Dynamic loading of packages

### ISSUE THREAD

**Issue by mitar** [meteor/meteor#6203]
Followup from #5714.

I am playing with a simple way for dynamic loading of packages. I changed file headers to be generated like:

``` javascript
(function (f) {
  if (typeof dependenciesReady !== 'undefined') {
    dependenciesReady(["packages/meteor.js","packages/underscore.js","packages/ejson.js"]).then(f);
  } else {
    f();
  };
})(function () {

/* Imports */

/* Package-scope variables */

// ...

});
```

And `dependenciesReady` is a simple promise which checks if `<script>` exists and if not it adds it:

``` javascript
function requireDependency(requirements) {
  if (requirements.length === 0) {
    return Promise.resolve();
  }

  var path = requirements[0];
  requirements = requirements.slice(1);

  if (Object.prototype.toString.call(path) === '[object Array]') {
    return requireDependency(path).then(function () {
      return requireDependency(requirements);
    })
  }

  var dependencyUrl = '/' + path;
  var dependencyElement = document.querySelector('script[src="' + dependencyUrl + '"]');
  if (dependencyElement) {
    // Our script tag, we added "ready" promise. Let us use it.
    if (dependencyElement.ready) {
      return dependencyElement.ready.then(function () {
        return requireDependency(requirements);
      });
    }
    else {
      // Not our script tag. Let us assume it is already loaded (probably
      // only tag for this code here is like this, which we are already
      // running, so it must have loaded).
      return requireDependency(requirements);
    }
  }

  var script = document.createElement('script');
  script.type = 'text/javascript';
  script.async = false;
  script.defer = false;
  script.ready = new Promise(function (resolve, reject) {
    script.onload = function (event) {
      resolve();
    };
    script.onerror = function (event) {
      reject(new Error("Script '" + event.target.src + "' failed to load."));
    };
  });

  var head = document.head || document.getElementsByTagName('head')[0];
  head.appendChild(script);
  script.src = dependencyUrl;

  return script.ready.then(function () {
    return requireDependency(requirements);
  });
}

dependenciesReady = function dependenciesReady(/* args */) {
  var args = Array.prototype.slice.call(arguments, 0);
  return requireDependency(args);
};
```

So this code is the only code which has to be deployed to the client. Everything else is then dynamically resolved.

**Comment by martijnwalraven** [comment]
Could you add a short description of your use case, i.e. what problems you are trying to solve with this?

**Comment by mitar** [comment]
I am trying to solve that not all JavaScript is loaded immediately.

**Comment by mitar** [comment]
I think this has been implemented in Meteor 1.5?
